### PR TITLE
sync new changes in zos3

### PR DESCRIPTION
**Pull request by Omarabdul3ziz**
Related Prs on zos repo:
- https://github.com/threefoldtech/zos/pull/2427
- https://github.com/threefoldtech/zos/pull/2424
- https://github.com/threefoldtech/zos/pull/2412

### CODE REVIEW

**Review by muhamadazmy — COMMENTED** [review]
Is this to bring fixes from zos3 to zos4 ? you probably need to find a better way to do this in the future. From what I see it looks like zos3 is not going anywhere anytime soon. 

With careful design, maybe the zos3 code base can be better separated into smaller reusable packages. Which can be used by both zos3 and 4. 

Only specific version code can be different. 

Of course this is easier said than done.

File: pkg/registrar/registrar.go [hunk 1]
```diff
@@ -146,6 +152,10 @@ func (r *Registrar) register(ctx context.Context, cl zbus.Client, env environmen
 			if err := r.reActivate(ctx, cl, env); err != nil {
 				log.Error().Err(err).Msg("failed to reactivate account")
 			}
+		case <-time.After(updateLocationInterval):
```
[the remark this answers is not in the document]
Comment (Omarabdul3ziz) [reply pkg/registrar/registrar.go]: i did on the zos3 pr https://github.com/threefoldtech/zos/pull/2427/commits/7e0ee7dcd139c7b3ae63089637c429bc04a1852e was waiting for an approve there to cherry-pick the commit here. done now

**Comment by Omarabdul3ziz** [comment]
yes, it is not the best way to do it i was discussing with Ashraf about it.  a suggestion for the coming changes was to open two PRs from the same branch against the two repos. 

but indeed having reusable packages is a cleaner option

**Comment by iwanbk** [comment]
> Is this to bring fixes from zos3 to zos4 ? you probably need to find a better way to do this in the future.

Yes, i do have the same concerns about this double repos to Ashraf & Jan.

> yes, it is not the best way to do it i was discussing with Ashraf about it. a suggestion for the coming changes was to open two PRs from the same branch against the two repos.

I argue that it is not "is not the best way to do",  but it is a bad way to do it.
DRY or avoid duplicated code is a well known way that we already do almost everyday.

> With careful design, maybe the zos3 code base can be better separated into smaller reusable packages. Which can be used by both zos3 and 4.

> Only specific version code can be different.

> Of course this is easier said than done.

Yes, we should do it.
It is not easy but better we start it sooner than later, before the things going wrong cc @ashraffouda @xmonader

**Comment by Omarabdul3ziz** [comment]
- cherry-picking the commit with other changes merged gone wrong. so i created another pr https://github.com/threefoldtech/zos4/pull/21
- also opened an issue for https://github.com/threefoldtech/zos/issues/2452

**Pull request closed without merging**